Cython/Compiler/ExprNodes.py

```diff
@@ -13878,6 +13893,98 @@ def analyse_type_annotation(self, env, assigned_value=None):
             warning(annotation.pos, "Unknown type declaration in annotation, ignoring")
         return base_type, arg_type
 
+
+class AssignmentExpressionRhsNode(ExprNode):
+    """
+    A very thin wrapper around an ExprNode. It gets its type and tempness from the right-hand side of
+    the AssignmentExpressionNode that it's linked to. At the analyse_types stage it converts itself
+    to a CloneNode
+
+    Attribute:
+    parent_expression  - AssignmentExpressionNode
+    """
+    subexprs = []
+
+    @property
+    def arg(self):
+        return self.parent_expression.rhs
+
+    @property
+    def type(self):
+        return self.arg.type
+
+    @property
+    def is_temp(self):
+        return self.arg.is_temp
+
+    def infer_type(self, env):
+        return self.arg.infer_type(env)
+
+    def analyse_types(self, env):
+        return CloneNode(self.arg)
+
+
+class AssignmentExpressionNode(ExprNode):
+    """
+    Also known as a named expression or the walrus operator
+
+    Arguments
+    lhs - NameNode - not stored directly as an attribute of the node
+    rhs - ExprNode
+
+    Attributes
+    lhs        - ExprNode (really should be a NameNode but this is validated later)
+    rhs        - ExprNode
+    assignment - SingleAssignmentNode
+    """
+    # subexprs and child_attrs are intentionally different here, because the assignment is not an expression
+    subexprs = ["rhs"]
+    child_attrs = ["assignment", "rhs"]
+
+    is_temp = False
+
+    def __init__(self, pos, lhs, rhs, **kwds):
+        super(AssignmentExpressionNode, self).__init__(pos, lhs=lhs, rhs=rhs, **kwds)
+        assign_expr_rhs = AssignmentExpressionRhsNode(rhs.pos, parent_expression=self)
+        self.assignment = SingleAssignmentNode(pos=pos, lhs=lhs,
+                                            rhs=assign_expr_rhs,
+                                            is_assignment_expression=True)
```

Comment:
Can't the `rhs` of the `SingleAssignmentNode` be a `CloneNode` right from the start? Why does it have to be special? AFAIU, all we need is a non-owned reference to the RHS node, which is in turn owned by the AssExpNode.

Reply:
Done.